### ISSUE TITLE
.family shouldn't rely on a Map cache but instead have providers override ==

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/river_pod#358]
Otherwise this can cause surprising behaviors when combining .autoDispose and .family:

```dart
final family = Provider.autoDispose.family(...);

final provider = family(42);

<wait for the provider to be disposed>

final provider2 = family(42);

print(provider == provider2); // false
```

**Comment by rrousselGit** [comment]
This is usually not problematic

The main scenario where this behavior is surprising is when reusing a provider override:

```dart
final family = Provider.autoDispose.family(...);

final override = family(42).overrideWithValue(Value());

await tester.pump(ProviderScope(
  overrides: [override],
));

<wait for the provider to be disposed>

await tester.pump(ProviderScope(
  // the override doesn't work! Because the new pump relies on a different Provider instance
  // but the previous instance is overriden
  overrides: [override],
));

```

**Comment by rrousselGit** [comment]
Closing since this is fixed as part of 1.0.0-dev
